adapters/zmaticoo/zmaticoo.go

```diff
@@ -0,0 +1,164 @@
+	var adapterRequests []*adapters.RequestData
+	adapterRequest, errs := a.makeRequest(request)
+	if errs == nil {
+		adapterRequests = append(adapterRequests, adapterRequest)
+	}
+	return adapterRequests, errs
+}
+
+func (a *adapter) makeRequest(request *openrtb2.BidRequest) (*adapters.RequestData, []error) {
+	var errs []error
+	zmaticooExt, errs := getZmaticooExt(request)
+	if zmaticooExt == nil {
+		return nil, errs
+	}
+	err := transform(request)
+	if err != nil {
+		return nil, append(errs, err)
+	}
+	reqBody, err := json.Marshal(request)
+	if err != nil {
+		return nil, append(errs, err)
+	}
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    reqBody,
+		Headers: headers,
+	}, errs
+}
+
+func transform(request *openrtb2.BidRequest) error {
+	for i, imp := range request.Imp {
+		if imp.Native != nil {
+			var nativeRequest map[string]interface{}
+			nativeCopyRequest := make(map[string]interface{})
+			err := json.Unmarshal([]byte(request.Imp[i].Native.Request), &nativeRequest)
+			//just ignore the bad native request
+			if err == nil {
+				_, exists := nativeRequest["native"]
+				if exists {
+					continue
+				}
+				nativeCopyRequest["native"] = nativeRequest
+				nativeReqByte, err := json.Marshal(nativeCopyRequest)
+				//just ignore the bad native request
+				if err != nil {
+					return err
+				}
+				nativeCopy := *request.Imp[i].Native
+				nativeCopy.Request = string(nativeReqByte)
+				request.Imp[i].Native = &nativeCopy
+			} else {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func getZmaticooExt(request *openrtb2.BidRequest) (*openrtb_ext.ExtImpZmaticoo, []error) {
+	var extImpZmaticoo openrtb_ext.ExtImpZmaticoo
+	var errs []error
+	for _, imp := range request.Imp {
+		var extBidder adapters.ExtImpBidder
+		err := json.Unmarshal(imp.Ext, &extBidder)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		err = json.Unmarshal(extBidder.Bidder, &extImpZmaticoo)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		break
+	}
+	return &extImpZmaticoo, errs
+
+}
+
+// MakeBids make the bids for the bid response.
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if adapters.IsResponseStatusCodeNoContent(response) {
+		return nil, nil
+	}
+	if err := adapters.CheckResponseStatusCodeForErrors(response); err != nil {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d.", response.StatusCode),
+		}}
+	}
+	var bidResp openrtb2.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+	var errs []error
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
```

Comment:
Initialising bidResponse with `1` is not going to give any performance benefit. Requesting you to please consider changing it to `adapters.NewBidderResponseWithBidsCapacity(len(internalRequest.Imp))`

Reply:
ok·